final/master.py

```diff
@@ -0,0 +1,408 @@
+
+                    sum_of_squares = math.sqrt(math.pow(current_red, 2) +
+                                               math.pow(current_green, 2) +
+                                               math.pow(current_blue, 2))
+
+                    # allows changing of the comparison parameters on successive executions
+                    sum_of_squares *= switch_comparison_direction
+                    sum_of_squares_average *= switch_comparison_direction
+
+                    # If pixel above or below the average, do something interesting
+                    if sum_of_squares > sum_of_squares_average:
+
+                        # Create a tail fading to black from that pixel
+                        for i in range (0, tail_length):
+                            if x - i > 0:
+                                new_red =  current_red - (current_red * i / tail_length)
+                                new_green = current_green - (current_green * i / tail_length)
+                                new_blue = current_blue - (current_blue * i / tail_length)
+
+                                pixel_array[x - i, y] = (new_red, new_green, new_blue)
+                                ignore_next_batch = True
+
+                    else:
+                        # Completes the effect of tail fading to black, by setting all else to black,
+                        pixel_array[x, y] = BLACK
+
+        # Draw new picture from altered pixel_array
+        pict = pixel_array.make_surface()
+        del pixel_array
+        del px_array
+
+        window.blit(pict, (0, 0))
+        pygame.display.update()
+
+        print ('Done, press space then i again if it is mostly black.')
+        switch_comparison_direction *= -1
+        print_controls()
+    else:
+        print ('It really will turn black if you run it again.')
+        print ('press space to reset.')
+        print_controls()
+
+
+def simplify_colour():
+
+    # The step in the colour range which each colour value will be rounded to
+    simplify_strength = 255 / 3
+
+    for x in range(0, WIDTH - 1):
+        for y in range(0, HEIGHT - 1):
+            red = window.get_at((x, y)).r
+            green = window.get_at((x, y)).g
+            blue = window.get_at((x, y)).b
+
+            # Essentially this divides by a number, truncates the decimal portion then multiplies it back to scale
+            # for instance were the strength = 100, a value of 234 would go to 2.34 to 2 then return 200
+            new_red = clamp(simplify_strength * int(red / simplify_strength) + (simplify_strength / 2))
+            new_green = clamp(simplify_strength * int(green / simplify_strength) + (simplify_strength / 2))
+            new_blue = clamp(simplify_strength * int(blue / simplify_strength) + (simplify_strength / 2))
+
+            px_array[x, y] = (new_red, new_green, new_blue)
+
+
+def pixelise_is_a_word_stop_underlining_it_in_grey():
+
+    tile_size = 20
+    tile_spacing = 5
+    global px_array
```

Comment:
[`global` considered harmful!](http://wiki.c2.com/?GlobalVariablesConsideredHarmful) --- the use of `global` should be avoided wherever possible for the reasons outlined in the article. Its use, restricted to the most esoteric of contexts, where it is the only appropriate option. Consider, instead, passing the variable `px_array` as an argument into the function. For example:

```
def pixelise_is_a_word_stop_underlining_it_in_grey(px_array):
    # body of function goes here
```